Do not install environment variables

haampie opened the pull request:
Edit:

Instead of installing the full `env` to `<install prefix>/.spack/spack-build-env.txt` only install variables that have been changed in the build environment. This is to avoid "installing" secret tokens, passwords and keys often set in CI.

Halfway this PR the original behavior of installing all variables was still available through `spack install --env-variables`, but that flag is now dropped (67e2bea01dd04c07161c53f587440071fd2c6f38).

<details>
 <summary>Original description:</summary>

We shouldn't persist environment variables to disk by default, and we should definitely not install them on shared instances and pipelines.

This PR ensures that we do not have a `<install prefix>/.spack/spack-build-env.txt` file and that the `<stage dir>/spack-build-env.txt` file has 600 permissions (that is read-write for the user only: `-rw-------`)

The only potentially safe thing to do is to store the environment *modifications* of spack's build environment, but until that is figured out, it's better to drop this entirely.
</details>

haampie commented:
@tgamblin this is ready now.

Pinging @vsoch: you may want to try the `environment_variables` analyzer after this PR, I think by default it just outputs nothing when spack-build-env.txt is missing in the install dir.

vsoch commented:
@haampie could you please add a warning that the file does not exist?

haampie commented:
Done @vsoch

Then this should be ready.

vsoch commented on lib/spack/spack/analyzers/environment_variables.py:
```diff
@@ -43,6 +44,7 @@ def _read_environment_file(self, filename):
         """
         if not os.path.exists(filename):
+            tty.warn("No environment file available")
             return
```
Returning None here will  be passed into the results dict and then not trigger the if statement here: https://github.com/spack/spack/blob/b83f06df0c4262afd43bf8816bae0db84247f849/lib/spack/spack/analyzers/analyzer_base.py#L99. So that should be good!

vsoch commented on lib/spack/spack/package.py:
```diff
@@ -1710,6 +1710,10 @@ def do_install(self, **kwargs):
             force (bool): Install again, even if already installed.
             install_deps (bool): Install dependencies before installing this
                 package
+            install_env_variables (bool): By default, the build environment
```
I'm glad this will be the default - it's a bit unexpected for a tool to save your entire environment without you really explicitly asking for it.

vsoch commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -176,6 +176,19 @@ def test_install_with_source(
+def test_install_env_variables(
+    flag, has_file, mock_packages, mock_archive, mock_fetch, config, install_mockery
+):
+    spec = Spec('libdwarf')
```
Can we have a test that if the flag isn't there the file does not exist?

haampie replied:
It's parametrized and testing that too :)

vsoch replied:
Totally missed that, thanks!

adamjstewart reviewed (changes requested):
> The only potentially safe thing to do is to store the environment modifications of spack's build environment, but until that is figured out, it's better to drop this entirely.

I agree that it's not a good idea to save _all_ env vars. Unfortunately, a lot of packages like PyTorch are configured entirely through env vars, so these are very important for reporting build issues. It should actually be very easy to only save env modifications, right @alalazo? We already have an env modifications object that is returned by `setup_build_environment`, we just need to save that to disk.

haampie commented:
> It should actually be very easy to only save env modifications

Sure, it's not hard, but this should also include some refactoring, cause env modifications are still coupled to `os.environ`, for example:

```
In [1]: m = spack.util.environment.EnvironmentModifications()

In [2]: m.set("A", "")

In [3]: m.unset("B")

In [4]: m.shell_modifications()
Out[4]: "export A='';\n"
```

Only when the environment variable B exists it'll show the `unset` -- this is not very useful for debugging...

```
In [5]: os.environ["B"] = "something"

In [6]: print(m.shell_modifications())
export A='';
unset B;
```

I'd rather fix this in a follow up PR.

adamjstewart commented:
Not displaying env vars that get unset is a minor issue in my opinion. In my experience, things left for future PRs never actually get done.

haampie commented:
Hm, okay... Do note that spack-env.txt is still in the staging dir, it's just not installed.

haampie commented:
@adamjstewart if I implement what you suggest (install the env changes), should we keep or drop the `spack install --env-variables` flag introduced in this PR? If we keep it, should I introduce another file `spack-env-mods.txt` that is installed unconditionally?

adamjstewart commented:
Personally I would drop it. But I would wait and see if others have opinions.

P.S. I didn't realize we were still generating the `spack-build-env.txt` file, just not installing it. So that makes some of my above points about debugging failed installations obsolete.

haampie commented:
I would be in favor of dropping the flag too and install the env modifications file unconditionally as spack-build-env.txt to keep things simple.

haampie commented:
@adamjstewart can you give the last two commits a try?

haampie commented on lib/spack/spack/util/environment.py:
```diff
@@ -968,7 +973,7 @@ def _source_single_file(file_and_args, environment):
         # Python interpreter, not necessarily sys.executable.
         python_cmd = executable.which('python3', 'python', 'python2')
-        python_cmd = python_cmd.name if python_cmd else sys.executable
+        python_cmd = python_cmd.path if python_cmd else sys.executable
```
This is not strictly necessary after 9ace7f1f3ba24833dcc5dc0d7e91daa0941a3161, but below we're sourcing a file, which may change PATH, so it's better to work with absolute paths to than executable names...

adamjstewart commented:
I can test this later tonight or this weekend. Ping me if I forget. Internship is keeping me busy during the day.

vsoch commented on lib/spack/spack/build_environment.py:
```diff
@@ -792,43 +795,45 @@ def load_external_modules(pkg):
 
-        if (not dirty) and (not env.is_unset('CPATH')):
+        if (not dirty) and (not env_mods.is_unset('CPATH')):
             tty.debug("A dependency has updated CPATH, this may lead pkg-"
```
should this be debug or perhaps a warning? I'm just thinking probably most people don't run things in debug. I know it was already here, but since it's part of the PR it's worth bringing up for discussion.

vsoch commented on lib/spack/spack/build_environment.py:
```diff
@@ -861,12 +866,16 @@ def setup_package(pkg, dirty, context='build'):
+    validate(env_mods, tty.warn)
+
+    # Compose and apply the full environment modifications
+    env.extend(env_mods)
```
I'm probably not familiar with this workflow - but if we are extending the clean env to include the mods ("not clean") what is the point of keeping the two separate? I guess I thought the point was to possibly save a cleaner version to file. So I think my questions are:

1. where does this returned environment (with both) go?
2. Is the one saved passed to `pkg.architecture.platform.setup_platform_environment(pkg, env_mods)` so it's ok?

haampie replied:
There is one base envmods (the clean env, or empty when --dirty), and one compiler/package modifications envmods. This has to be split, because we validate the package envmods for suspicious changes such as setting the compiler variable etc. Only after validation these modifications are merged, and we log all the modifications together.

vsoch commented on lib/spack/spack/installer.py:
```diff
@@ -547,8 +547,8 @@ def log(pkg):
-    # Archive the environment used for the build
-    fs.install(pkg.env_path, pkg.install_env_path)
+    # Archive the environment modifications for the build.
+    fs.install(pkg.env_mods_path, pkg.install_env_path)
```
okay maybe this answers my question if this pkg.env_mods_path is the one we saw passed earlier (the clean one)

vsoch commented on lib/spack/spack/installer.py:
```diff
@@ -1683,6 +1683,10 @@ def build_process(pkg, kwargs):
     unmodified_env = kwargs.get('unmodified_env', {})
+    env_mods = kwargs.get(
+        'env_modifications',
+        spack.util.environment.EnvironmentModifications()
```
why not the clean one here?

haampie replied:
not sure what you mean

vsoch reviewed (commented):
I'm not sure I'm super useful as a reviewer, but I've asked questions that maybe others have that might be helpful for the PR.

haampie committed:
Only install env modifications in <prefix>/.spack

- Do not store the full list of environment variables in
  <prefix>/.spack/spack-build-env.txt because it may contain user secrets.

- Only store environment variable modifications upon installation.

- Variables like PATH may still contain user and system paths to make
  spack-build-env.txt sourceable. Variables containing paths are
  modified through prepending/appending, and if we don't apply these
  to the current environment variable, we end up with statements like
  `export PATH=/path/to/spack/bin` with system paths missing, meaning
  no system binaries are in the path, which is a bad user experience.

- Do write the full environment to spack-build-env.txt in the staging dir,
  but ensure it is readonly for the current user, to make it a bit safer
  on shared systems.

haampie commented:
Finally rebased this.

```
- Do not store the full list of environment variables in
  <prefix>/.spack/spack-build-env.txt because it may contain user secrets.

- Only store environment variable modifications upon installation.

- Variables like PATH may still contain user and system paths to make
  spack-build-env.txt sourceable. Variables containing paths are
  modified through prepending/appending, and if we don't apply these
  to the current environment variable, we end up with statements like
  `export PATH=/path/to/spack/bin` with system paths missing, meaning
  no system binaries are in the path, which is a bad user experience.

- Do write the full environment to spack-build-env.txt in the staging dir,
  but ensure it is readonly for the current user, to make it a bit safer
  on shared systems.
```

becker33 commented on lib/spack/spack/util/environment.py:
```diff
@@ -593,21 +594,26 @@ def apply_modifications(self, env=None):
             new = new_env.get(name, None)
-            old = os.environ.get(name, None)
-            if new != old:
+            old = env.get(name, None)
+            if explicit or new != old:
```
This feels a little hacky, we should probably push the shell modifications information into the modifications objects. Something like

```
class SetPath(...):
    ...
    def shell_modifications(self, shell='sh'):
        if shell == 'sh':
            return 'export {0}={1};'.format(self.name, self.value)
       ...
...

class PrependPath(...):
    ...
    def shell_modifications(self, shell='sh'):
        if shell = 'sh':
            return 'if [[ -z ${0} ]]; then export {0}={1}:${0}; else; export {0}={1}; fi;'.format(self.name, self.value)
        ...
```

haampie replied:
Something along those lines would be interesting because it could hide even more of the user environment: `export PATH="/spack/prefix/bin:$PATH"` instead of `export PATH="/spack/prefix/bin:/home/my-full-name/local/bin"`

But I think it's important that `source spack-env-mods.txt` is idempotent so running it twice has no effects (you might actually run it after some debugging of a failing build inside build-env to reset variables), and `export x="[something-by-spack]:$x`" is not.

Also, I like that spack applies the *effective* changes in one go instead of each individual modification -- in particular with the changes I've made above, you even get a nice alphabetically sorted list, so it's very easy to find a variable and see what it's actual value is. If we would want to keep that behavior, I'm not sure how that would be supported by the above (diffing the applied changes with from_environment_diff?)

Finally, the `explicit` stuff I've introduced is just to also output variables that are touched by Spack, but are effectively unchanged. For instance if you've `spack load`'ed a dependency of a spec you then `spack install`, and this dependency sets `PYTHONHOME` to its prefix, then I think spack wouldn't log it, because the build will load the same dependency and `PYTHONHOME` is effectively unchanged (whether its cleared in clean_environment() or not, it is effectively unchanged).

haampie replied:
any thoughts @becker33?

becker33 reviewed (commented):
Mostly lgtm

tgamblin commented:
@haampie this needs a rebase

haampie commented:
done

haampie commented:
and again...

haampie commented:
and again

sethrj commented:
Needs reapproval from @adamjstewart  but it looks good to me and it seems from reading the discussion that all the major points have been agreed upon.
